packages/@react-spectrum/s2/src/Calendar.tsx

```diff
@@ -10,38 +10,585 @@
+
+  // Starting from the current day, find the first day before it in the current week that is not selected.
+  // Then, the span of selected days is the current day minus the first unselected day.
+  let firstUnselectedInRangeInWeek = datesInWeek.slice(0, dayIndex + 1).reverse().findIndex((date, i) => {
+    console.log('date', date?.month === props.date.month, date?.month, props.date.month);
+    return date && i > 0 && (!state.isSelected(date) || date.month !== props.date.month);
+  });
+  let selectionSpan = -1;
+  if (firstUnselectedInRangeInWeek > -1 && isSelected) {
+    selectionSpan = firstUnselectedInRangeInWeek - 1;
+  } else if (isSelected) {
+    selectionSpan = dayIndex;
+  }
+
+  let isBackgroundStyleApplied = (
+    isSelected
+    && (isLastChild || !isNextDaySelected)
+    && isRangeSelection
+    && (state.isSelected(date.subtract({days: 1}))
+      || state.isSelected(date.add({days: 1}))
+    ));
+
+  return (
+    <div
+      className={style({
+        position: 'relative',
+        width: 32,
+        '--cell-width': {
+          type: 'width',
+          value: '[self(width)]'
+        }
+      })}>
+      <div
+        ref={ref}
+        style={pressScale(ref, {})(renderProps!)}
+        className={cellInnerStyles({...renderProps!, selectionMode: isRangeSelection ? 'range' : 'single'})}>
+        <div>
+          {formattedDate}
+        </div>
+        {isUnavailable && <div className={unavailableStyles} role="presentation" />}
+      </div>
+      {isBackgroundStyleApplied && <div style={{'--selection-span': selectionSpan} as CSSProperties} className={selectionSpanStyles} role="presentation" />}
+    </div>
+  );
+};
+
+type DayOfWeek = 'sun' | 'mon' | 'tue' | 'wed' | 'thu' | 'fri' | 'sat';
+
+/**
+ * Calculate the week index (0-based) and day index (0-based) for a given date within a month in a calendar.
+ * @param date - The date to calculate indices for.
+ * @param locale - The locale string (e.g., 'en-US', 'fr-FR', 'hi-IN-u-ca-indian').
+ * @param firstDayOfWeek - Optional override for the first day of the week ('sun', 'mon', 'tue', etc.).
+ * @returns Object with weekIndex and dayIndex.
+ */
+function useWeekAndDayIndices(
+  date: CalendarDate,
+  locale: string,
+  firstDayOfWeek?: DayOfWeek
+) {
+  let {dayIndex, weekIndex} = useMemo(() => {
+    // Get the day index within the week (0-6)
+    const dayIndex = getDayOfWeek(date, locale, firstDayOfWeek);
+
+    const monthStart = startOfMonth(date);
+
+    // Calculate the week index by finding which week this date falls into
+    // within the month's calendar grid
+    const monthStartDayOfWeek = getDayOfWeek(monthStart, locale, firstDayOfWeek);
+    const dayOfMonth = date.day;
+
+    const weekIndex = Math.floor((dayOfMonth + monthStartDayOfWeek - 1) / 7);
+
+    return {
+      weekIndex,
+      dayIndex
+    };
+  }, [date, locale, firstDayOfWeek]);
+
+  return {dayIndex, weekIndex};
+}
+
+function useDefaultFirstDayOfWeek(locale: string): DayOfWeek {
```

Comment:
hmm doesn't `@internationalized/date` do this automatically if the `firstDayOfWeek` option is not passed into its functions?